### DISCUSSION TITLE
Send() doesn't work when invoked callback handler is async

### DISCUSSION THREAD

**Discussion by moon0326** [statelyai/xstate#4045] [Q&A] [answered]
Hi ✋ ,

I have an async invoked callback handler as below:
```
invoke: {
  id: "ponger",
  src: (context, event) => async (callback, onReceive) => {
    // Whenever parent sends 'PING',
    // send parent 'PONG' event
    onReceive((e) => {
      console.log(e);
    });
  }
},
```

I noticed that an asynchronous handler can't receive events.

Here's the CodeSandbox link: https://codesandbox.io/s/xstate-example-template-vanilla-forked-mh1pis?file=/src/index.js

It's essentially the same as the example from https://xstate.js.org/docs/guides/communication.html#invoking-callbacks, but with an asynchronous approach.

I must be missing something or misusing the callback.

Please advise 🙏

**Comment by Andarist** [comment, marked as answer]
Please don't make that callback `async`. Currently, we create its result as a promise actor but notice that your `async` function completes immediately - that means that we complete that promise actor internally and we close the communication channel between it and the parent actor. 

If you need to use `async` function there for convenience then u can introduce an async IIFE like this:
```ts
        src: (context, event) => (callback, onReceive) => {
          (async () => {
            // Whenever parent sends 'PING',
            // send parent 'PONG' event
            onReceive((e) => {
              console.log(e);
            });
          })();
        }
```

**Reply by moon0326** [reply]
Thank you so much 👍
